Classes/Core/Functional/FunctionalTestCase.php

```diff
@@ -1034,47 +1035,54 @@ private function retrieveFrontendSubRequestResult(
         // carry that information.
+        // @todo This global can be removed, after Instructions are handled using the ServerRequest
+        //       directly instead of loading the RequestBootstrap and thus this global handover.
         $_SERVER['X_TYPO3_TESTING_FRAMEWORK']['request'] = $request;
 
-        // Create ServerRequest from testing-framework InternalRequest object
-        $uri = $request->getUri();
-
-        // Build minimal serverParams and hand over to ServerRequest. The normalizedParams
-        // attribute relies on these. Note the access to $_SERVER should be dropped when the
-        // above getIndpEnv() can be dropped, too.
-        $serverParams = [
-            'SCRIPT_NAME' => $_SERVER['SCRIPT_NAME'],
-            'HTTP_HOST'  => $_SERVER['HTTP_HOST'],
-            'SERVER_NAME' => $_SERVER['SERVER_NAME'],
-            'HTTPS' => $uri->getScheme() === 'https' ? 'on' : 'off',
-            'REMOTE_ADDR' => '127.0.0.1',
-        ];
+        /** @var InternalRequest $serverRequest */
+        $serverRequest = $request->withAttribute('typo3.testing.context', $context);
+
+        // If no serverParams are set, fallback to to old ServerRequest build behaviour and
+        // fill needed stuff. This has been historically be done to create ServerRequest from
+        // InternalRequest which is now directly used.
+        if ($serverRequest->getServerParams() === []) {
+            $serverRequest = $serverRequest
+                ->withServerParams(
+                    [
+                        'SCRIPT_NAME' => '/index.php',
+                        'HTTP_HOST'  => $serverRequest->getUri()->getHost(),
+                        'SERVER_NAME' => $serverRequest->getUri()->getHost(),
+                        'HTTPS' => $serverRequest->getUri()->getScheme() === 'https' ? 'on' : 'off',
+                        'REMOTE_ADDR' => '127.0.0.1',
+                    ]
+                );
+        }
 
-        $serverRequest = new ServerRequest(
-            $uri,
-            $request->getMethod(),
-            $request->getBody(),
-            $request->getHeaders(),
-            $serverParams
-        );
-        $serverRequest = $serverRequest->withAttribute('typo3.testing.context', $context);
-        $requestUrlParts = [];
-        parse_str($uri->getQuery(), $requestUrlParts);
-        $serverRequest = $serverRequest->withQueryParams($requestUrlParts);
-        $parsedBody = $request->getParsedBody();
-        if (empty($parsedBody) && $request->getBody() !== null && in_array($request->getMethod(), ['PUT', 'PATCH', 'DELETE'])) {
-            parse_str((string)$request->getBody(), $parsedBody);
+        // Ensure that query parameters in request uri and parsed queryParameters are in sync.
+        $queryParams = [];
+        parse_str($serverRequest->getUri()->getQuery(), $queryParams);
+        $serverRequest = $serverRequest->withQueryParams($queryParams);
```

Comment:
the query logic should be similar to the body logic below.

Reply:
Done. Also added test coverage for this to core request tests:
https://review.typo3.org/c/Packages/TYPO3.CMS/+/73990